### DISCUSSION TITLE
Parse JSON directly from decimal uint8 vector

### DISCUSSION THREAD

**Discussion by NicolaBlasuttigh** [nlohmann/json#2321] [Q&A] [answered]
Hi!
I am working in Simulink on a TCP connection between client (my computer) and server. Through standard Ethernet blocks, I send an HTTP request like the following:

```
GET /status HTTP/1.1
Host:...
Accept: application/json
```
To send it, I convert it with the `uint8('GET /status...')` command in Simulink. The server responds, always with a uint8 array (decimal bytes) like this :

`resp=[72 84 84 80 47 49 32 50 48...]       //(HTTP/1.1 200 OK... ).`

In the reply "resp" content, always in decimal characters, **there is a text in JSON** format like {"VarA":1000, "VarB":2000,...}.

My purpose is to create an S-function (but more generally a code in C++) that taking in input the vector "resp", returns the values of VarA and VarB (1000 and 2000).

My idea was to convert "resp" into a string and then pass it to the json:parse functions. Is this correct? Can it be done?

More important: I also wonder if it is right to convert it into a string and then work on it...does anyone know if json::parse() can work directly on the uint_8 vector without the string conversion?

Thank you everyone!

**Comment by nlohmann** [comment, marked as answer]
You can pass an iterator range:

```cpp
#include <iostream>
#include "json.hpp"

using json = nlohmann::json;

int main()
{
    std::vector<std::uint8_t> resp = {'f', 'o', 'o', '{', '"', 'V', 'a', 'r', '"', ':', '1', '0', '}'};
    json j = json::parse(resp.begin() + 3, resp.end());
    std::cout << j << std::endl;
}
```

Output:

```json
{"Var":10}
```

**Reply by NicolaBlasuttigh** [reply]
Thank you @nlohmann ! I hadn't seen this solution! Anyway, the problem is that your resp vector is a character vector..right? I have a decimal uint8_t vector instead. I should convert it to a character vector and then pass it to parse(). I wonder if to convert decimal to string is a good idea (or the only one) or I can pass directly the decimal vector. This last question because I need to keep the computational cost to a minimum.
Thank you!

**Reply by nlohmann** [reply]
Vector `resp` in my example is of type `std::vector<std::uint8_t>`. I just initialized the values using character literals to make clear what's inside. The individual values are still `std::uint8_t`. In any case, the library can parse from any container with 1, 2, or 4 bit values inside, interpreting them as UTF-8, UTF-16, and UTF-32, respectively.

**Reply by NicolaBlasuttigh** [reply]
Oh yes, right! Thank you! Still learning...

And what if "resp" isn't really a vector but a pointer? Let me get this straight.

In Simulink S-functions I initialize a pointer like this:

 ```
   InputPtrsType u = ssGetInputPortSignalPtrs(S,0); 
   InputUInt8PtrsType resp8 = (InputUInt8PtrsType)u; 
```

where resp8 now contains all the response.

If I write:

`std::cout << *resp8 << std:endl`

Print the whole answer. (HTTP 200... {"Var":10, "Var2:...})

But if I write:
`json j = json::parse(*resp8);`
an error comes up:
_Unexpected exception ([json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - invalid literal; last read: 'H') was caught._ 

Probably because I'm also passing headers and I should start from "{".
I tried to see if it's possible to convert resp8 to an std::vector so I could use .begin() and end() but I don't understand if it's legal or feasible.

Could you help me on this too?
Thank you very much.

**Reply by nlohmann** [reply]
You can still pass two iterators, which would be pointers in your case. For the example above:

```cpp
json j = json::parse(resp.data() + 3, resp.data() + resp.size());
```

Here, we take the `std::uint8_t*` array via `resp.data()` and use pointer arithmetic to access the 4th value to start. The second pointer  `resp.data() + resp.size()` points to one value after the last one. If you know the size of your `resp8` array, then you can do the same, assuming you know when the headers end.

**Reply by NicolaBlasuttigh** [reply]
Thank you very much for your reply @nlohmann ! You really thought of everything! 

However, it keeps giving me error... the problem is that Simulink creates an `InputUInt8PtrsType` pointer rather than `std::uint8_t`.
The function data() belongs to `std::vector` so I can't use it with my pointer type.
However, I "solved" (I don't think in the most elegant way) by creating an std::vector and copying all the values of the pointer inside it.
In this way:

```
InputPtrsType u = ssGetInputPortSignalPtrs(S,0); 
InputUInt8PtrsType resp8 = (InputUInt8PtrsType)u; 

std::vector<std::uint8_t> resp_vec; 
    
for ( int i=0; i<=3000; i++ ){
     resp_vec.push_back(*resp8[i]); 
    }
```

Now I have my resp_vec std::vector which I can work with using `begin()`, `end()` and everything else.
It's probably possible to convert the pointer from one type to another but I don't know how and I've done it this way.

**About the end of the headers**, they always end with `\r\n\r\n (13 10 13 10 in decimal)`. Do you think it's possible to find those 4 bytes and start the parse() from the next number? (which will be "{" (123). It seems laborious but it's the only way to know where they end up since the headers could change their size! I was hoping to look for the first "{" but the cookie also uses "{".

Thank you again!

**Reply by nlohmann** [reply]
- Did you try passing the pointers of type `InputUInt8PtrsType` directly to `parse`? Making a copy seems so wasteful.
- I'm afraid there is no function in the library to skip until `\r\n\r\n` is read. You could write a function that just advances a pointer until you see the pattern and then pass that pointer as first argument to `parse`.

**Reply by NicolaBlasuttigh** [reply]
- Yes! That's exactly what I did and wrote before!
> But if I write:
> json j = json::parse(*resp8);
> an error comes up:
> Unexpected exception ([json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - invalid literal; last read: 'H') was caught.

And if I write ` json j = json::parse(resp8);` 
**Building Error** _Reason:
impossible to convert from 'InputUInt8PtrsType' to 'nlohmann::detail::input_adapter'._

Actually,` json j = json::parse(*resp8);` would work, but given that it's not a std::vector type I cannot use begin(), end(), etc... And I didn't find similar functions to work with pointers!
Anyway, you gave me a lot of help! I think that when I'll move everything in C++ (and not in S-function) will be easier!
- Fine! I'll try to write something and then I'll share here!

**Reply by nlohmann** [reply]
What I meant is not passing a single pointer, but two pointer like I described in https://github.com/nlohmann/json/discussions/2321#discussioncomment-43849. You do not need `begin()`/`end()`, but can use offsets. Something like

```cpp
json j = json::parse(resp8, resp8 + size);
```
